Canner/vulcan-sql
# Feature: Implementing restful route generator and api server according to api schema

## Description
This PR includes some components for two parts:
### 1.  Implementing validators from **IValidator**

- `IntegerTypeValidator`, `UUIDTypeValidator`, `StringTypeValidator` and `DateTypeValidator` which implementing **IValidator** interface to validate input data according validator fields of APISchema.

### 2. Getting API schemas and receiving requests to validate and respond to results:
- `RequestTransformer`: transform and check data type according to `type` in schema from koa context request.
- `RequestValidator`: validate transformed request according to `validators` in schema.
- `RestfulRoute`: handling the transformed request ( it extends `BaseRoute` which handles transformer and validator ) for restful API, the handler does not implement.
- `GraphQLRoute`: handling the transformed request ( it extends `BaseRoute` which handles transformer and validator ) graphQL, the handler does not implement.
- `RouteGenerator`: a factory pattern to create graphQL or restful route according to API type.
- `VulcanApplication`:  array of`GraphQLRoute` or `RestfulRoute` and run the server by `listen`.

Below is class relationship : 
<p align="center">
  <img src="https://user-images.githubusercontent.com/5389253/168508958-3a8367dc-9173-4b5d-975c-d7986214d818.png" />
</p>

## Finish Content for original Ivan drew architecture 
Below is the finished par which maps to original Ivan drew architecture:
![oss drawio](https://user-images.githubusercontent.com/5389253/168512701-e16d530c-0046-4a4d-9ffa-56414be5af38.png)


## How To Test / Expected Results
For the test result, please see the below test cases passed of unit test:
<img width="710" alt="Unit-test" src="https://user-images.githubusercontent.com/5389253/168511852-65c8d420-007c-4e19-85dc-099c393dfa95.png">

You could check `app.spec.ts` for knowing how to run the `VulcanApplication`.

## Commit Messages
- https://github.com/Canner/vulcan/pull/7/commits/44512c3f0d0cc2f481db01f0d2612d13df401549 - feature(core): add built-in data type validators and testcase 
    - add `DateTypeValidator`, `IntegerTypeValidator`, `StringTypeValidator`, `UUIDTypeValidator` and its test cases.
    - add `loadedValidators` for loading all built-in and user-defined validators dynamically.
    - install joi, moment plugins at package.json in the core folder.
    - install faker, ts-sinon, uuid plugins at vulcan package.json.

- https://github.com/Canner/vulcan/pull/7/commits/e5e8eef1891e854a5459cf4b5895a728707a65f5 - feature(serve): add request transformer, validator, and route generator to generate routes for handling and check request
    - add `ReuqestTransformer`, `RequestValidator`, for converting request data format and validating.
    - add `BaseRoute`, `RestfulRoute`, `GraphQLRoute` to create route.
    - add `RouteGenerator` to generate Route according to api route type.
    - add `VulcanApplication` to add routes and create the server.
    - add test cases for `ReuqestTransformer`, `RequestValidator`, `RouteGenerator` and `VulcanApplication`.
    - install `suepertest` in package.json for testing the koa server.
    - add `start` command to run serve package directly for testing in project.json.
    - update tsconfig.json for making alias path could be run by ts-node in nx.
